playback/src/player.rs

```diff
@@ -274,9 +278,15 @@ impl PlayerInternal {
 
             None => {
                 self.stop_sink();
-                self.run_onstop();
 
-                let old_state = mem::replace(&mut self.state, PlayerState::Stopped);
+                let new_state = match self.state {
```

Comment:
`handle_packet` should only be called from a `Playing` state, so you can `panic!` on other cases.

Additionally you can change `signal_end_of_track` into `playing_to_end_of_track(&mut self)`, which takes care of switching state and sending the signal.